text/0000-wire-reform.md

```diff
@@ -0,0 +1,197 @@
+* `@wire` decorator will create an instance of the provided `WireAdapter` during the component initialization routine by providing the data callback as argument.
+* `@wire` decorator will invoke the adapter's `connect` and `disconnect` based on the internal hooks used per component.
+* `@wire` decorator will create a mutation tracking phase to track any access executed during the computation of the config before calling the `update` routine to be able to detect mutations on those values and issue another update on the adapter instance.
+* `@wire` decorator will detect if the adapter is expecting contextual information, and if there is a context provider defined to carry on the hand-shaking protocol between the consumer and the provider.
+* `@lwc/compiler` will provide a config function per `@wire()` declaration to produce a new config object when invoked with the `component` as the first argument. The `@wire` adapter can rely on that config function to produce a new config object at will.
+* `@lwc/wire-service` becomes Lightning Platform-specific for the most part (`register()` method) since anyone can implement the wire adapter protocol.
+
+_Backwards Compatibility Notes:_
+
+* As today, the descriptor installed on the prototype of the component by the `@wire` decorator was identical to `track` when decorating a field. This means that the component author could change the value of the field, and such change will be tracked. This is not longer the case, and even though the author can still change the value, it will not be reactive, causing no side effects on the UI of the element. This is a non-backwards compatible change, but we believe this is a very low risk for something that was not working as expected.
+
+### Wire Adapter Protocol
+
+The formalization of the wire adapter protocol is important because that enables the interoperability aspect of this feature. The adapter's code should not be aware of the component system, or even the application framework. It only cares about very specific hints to produce a stream of data. The following describes the proposed protocol:
+
+```ts
+interface WireAdapter {
+    update(config: ConfigValue, context?: ContextValue);
+    connect();
+    disconnect();
+}
+interface WireAdapterConstructor {
+    new (callback: DataCallback): WireAdapter;
+    configSchema?: Record<string, WireAdapterSchemaValue>;
+    contextSchema?: Record<string, WireAdapterSchemaValue>;
+}
+type DataCallback = (value: any) => void;
+type ConfigValue = Record<string, any>;
+type ContextValue = Record<string, any>;
+type WireAdapterSchemaValue = 'optional' | 'required';
+```
+
+_Notes:_
+
+* Not all environments will support or need context (e.g.: preloading LDS data), but does supporting it can rely on the static field called `contextSchema` to provide the context value when available.
+* Some environments might choose to implement validation rules for `configSchema` and `contextSchema` to guarantee compliance.
```

Comment:
What is the advantage `configSchema` and `contextSchema` vs doing the validation by hand in the `update` callback?
It feels that we are adding higher-level abstractions to low-level APIs.